File: Stacks-and-Queues/queue.rb
module QueueModule
  class Node
    attr_accessor :value, :next
    def initialize(value)
      @value = value
      @next = nil
    end
  end

  class Queue
    attr_accessor :first, :last, :length
    def initialize(value)
      new_node = Node.new(value)
      @first = new_node
      @last = new_node
      @length = 1
    end
    def enqueue(value)
      new_node = Node.new(value)
      if(@length == 0)
        @first = new_node
        @last = new_node
      else
        @last.next = new_node
        @last = new_node
      end
      @length += 1
      return self
    end
    def dequeue
      if(@length == 0)
        return nil
      end
      temp = @first
      if(@length == 1)
        @first = nil
        @last = nil
      else
        @first = @first.next
        temp.next = nil
      end
      @length -= 1
      return temp
    end
  end
end

File: Stacks-and-Queues/queue.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Queue:
    def __init__(self, value):
        new_node = Node(value)
        self.first = new_node
        self.last = new_node
        self.length = 1

    def enqueue(self, value):
        new_node = Node(value)
        if(self.length == 0):
            self.first = new_node
            self.last = new_node
        else:
            self.last.next = new_node
            self.last = new_node

        self.length += 1
        return self

    def dequeue(self):
        if(self.length == 0):
            return None
        temp = this.first
        if(self.length == 1):
            self.first = None
            self.last = None
        else:
            self.first = self.first.next
            temp.next = None

        self.length -= 1
        return temp

